internal/implementations/implementation.go

```diff
@@ -0,0 +1,94 @@
+package implementations
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/events"
+	"github.com/nginxinc/nginx-kubernetes-gateway/pkg/sdk"
+)
+
+type ImplementationImpl struct {
+	logger       logr.Logger
+	objectType   client.Object
+	eventCh      chan<- interface{}
+	filter       NamespacedNameFilter
+	resourceKind string
+}
+
+var _ sdk.Implementation = &ImplementationImpl{}
+
+type NamespacedNameFilter func(nsname types.NamespacedName) (bool, string)
+
+func NewImplementation(objectType client.Object, logger logr.Logger, eventCh chan<- interface{}) *ImplementationImpl {
```

Comment:
Instead of creating a new generic-ish implementation object here, why not just modify the existing resource implementation to conform to a common interface?

```
type Implementation interface{
   Upsert(client.Object)
   Remove(types.NamespacedName)
}
```


 That way we don't need to do all of the type actions in order to figure out what resource we are working with.

Reply:
> Instead of creating a new generic-ish implementation object here, why not just modify the existing resource implementation to conform to a common interface?

yep. but those implementations will all look the same except for the resource type they deal with.